File: sdk/php/tests/FailureTest.php
<?php

declare(strict_types=1);

namespace Shojiku\Tests;

use PHPUnit\Framework\TestCase;
use Shojiku\Failure;
use Shojiku\Step;

final class FailureTest extends TestCase
{
    use EngineFixtures;

    public function testTheStepIsAlwaysThisPackagesOwnLifecycleStep(): void
    {
        // The engine's report names an INTERNAL stage in its own `step` field
        // — `render` for a refused document — and reading it here would make
        // the trace's step mean different things depending on which layer
        // refused.
        $failure = $this->client()->generate('broken', [])->failure();

        self::assertNotNull($failure);
        self::assertSame(Step::Generate, $failure->step());
        // What the engine said specifically is the kind.
        self::assertSame('document', $failure->kind());
    }

    public function testACauseChainFlattensOutermostFirst(): void
    {
        $inner = new Failure(Step::Sign, 'io', 'no such file');
        $outer = new Failure(Step::Sign, 'key', 'the key could not be loaded', [], $inner);

        self::assertSame([$outer, $inner], $outer->causes());
        self::assertSame($inner, $outer->cause());
        self::assertSame([$inner], $inner->causes());
    }

    public function testARejectedTemplateNameCarriesTheUnderlyingDetailAsItsCause(): void
    {
        $this->inTempDir(function (string $root): void {
            mkdir($root.'/empty');
            mkdir($root.'/empty/templates.yml');

            $failure = $this->client(['templates' => $root])->generate('empty', [])->failure();

            self::assertNotNull($failure);
            self::assertCount(2, $failure->causes());
            self::assertSame('io', $failure->causes()[1]->kind());
        });
    }

    public function testAFailurePrintsAsStepKindAndMessage(): void
    {
        $failure = new Failure(Step::Verify, 'signature', 'verification failed');

        self::assertSame('verify/signature: verification failed', (string) $failure);
        self::assertSame('verification failed', $failure->message());
        self::assertSame([], $failure->diagnostics());
    }
}

File: sdk/java/src/test/java/jp/kengos/shojiku/FailureTest.java
package jp.kengos.shojiku;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

import java.util.List;
import org.junit.jupiter.api.Test;

/** The failure trace: a value, with the chain underneath it inspectable rather than unwound. */
class FailureTest extends Fixtures {

  @Test
  void aFailureReadsItselfOffTheEnginesErrorPayload() {
    Failure failure =
        Failure.fromErrorJson(
            "{\"kind\":\"render_failed\",\"message\":\"no font\"}", Step.GENERATE, null, null);

    assertEquals(Step.GENERATE, failure.step());
    assertEquals("render_failed", failure.kind());
    assertEquals("no font", failure.message());
    assertNull(failure.cause());
    assertEquals(List.of(), failure.diagnostics());
  }

  @Test
  void anAbsentOrEmptyPayloadStillProducesAUsableFailure() {
    for (String payload : new String[] {null, ""}) {
      Failure failure = Failure.fromErrorJson(payload, Step.VERIFY, null, null);

      assertEquals("unknown", failure.kind());
      assertEquals("", failure.message());
    }
  }

  @Test
  void aPayloadMissingItsKeysFallsBackRatherThanThrowing() {
    Failure failure = Failure.fromErrorJson("{\"other\":1}", Step.SIGN, null, null);

    assertEquals("unknown", failure.kind());
    assertEquals("", failure.message());
  }

  @Test
  void causesFlattenTheChainOutermostFirst() {
    // What you log when you want the whole story rather than only its headline.
    Failure root = new Failure(Step.GENERATE, "io", "disk");
    Failure middle = new Failure(Step.GENERATE, "template_unreadable", "unreadable", null, root);
    Failure outer = new Failure(Step.GENERATE, "template_name", "refused", null, middle);

    assertEquals(
        List.of("refused", "unreadable", "disk"),
        outer.causes().stream().map(Failure::message).toList());
    assertEquals(1, root.causes().size());
  }

  @Test
  void aFailurePrintsItsStepAndKind() {
    assertEquals(
        "sign/key_unreadable: nope", new Failure(Step.SIGN, "key_unreadable", "nope").toString());
  }

  @Test
  void theStepVocabularyIsTheSdksOwnThreeAndNothingElse() {
    // The engine's error object names an INTERNAL stage (render, validate).
    // Passing it through would make this field mean different things depending on
    // which layer refused.
    assertEquals(
        List.of("generate", "sign", "verify"),
        java.util.Arrays.stream(Step.values()).map(Step::toString).toList());
  }
}
